Add ESMF v8.4.1, and number of other changes to improve ESMF integration with Spack

theurich opened the pull request:
This PR contains the following updates:

- Support for latest ESMF patch release v8.4.1.
- Update the ESMF package description to mention that NUOPC Layer is included.
- Update the ESMF web site to the official valid URL.
- Add two ESMF core team members to the list of maintainers.
- Indicate deprecated ESMF versions as such.
- Simplify/standardize variant selection wrt external vs internal ParallelIO usage via `external-parallelio`, in analogy to `external-lapack`.
- For ESMF >= 8.3.1 default `external-parallelio = True`.
- Provide override access to `ESMF_PIO` for rare cases where it is needed.
- Remove unnecessary required dependencies on "zlib" and "libxml2".
- Set minimum external ParallelIO version depending on ESMF version.
- Set `ESMF_CC` to a consistent C compiler in analogy to `ESMF_CXX` and `ESMF_F90`.
- Only use the `ESMF_F90COMPILEOPTS="-fallow-argument-mismatch"` work-around for older ESMF versions.
- Only use the `ESMF_COMM="mpich3"` setting for older ESMF version, otherwise use the more general `"mpich"` setting.

theurich committed:
Add v8.4.1, and a few other changes.

    Minor adjustments for better alignment between Spack and ESMF native
    build. For ESMF >= 8.3.1 now Spack defaults to using
    external-parallelio. Before use internal version, which was PIO-1 all
    the way up to v8.3.0b10 anyway! Xerces is disabled by default.

theurich commented:
@uturuncoglu, FYI.

theurich committed:
Deal with two long lines flagged by prechecks/style.

theurich committed:
Try to satisfy prechecks/style.

theurich committed:
Try to satisfy flake8 rules wrt indentation of continuation lines.

uturuncoglu commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -75,7 +63,9 @@ class Esmf(MakefilePackage):
     depends_on("parallel-netcdf@1.2.0:", when="+pnetcdf")
     depends_on("xerces-c@3.1.0:", when="+xerces")
-    depends_on("parallelio@2.5.8:", when="+parallelio")
+    depends_on("parallelio@2.5.7:", when="@8.3.0:8.3.99+external-parallelio")
```
It is good to have those. Thanks. It was fixed to 2.5.8 before for all the versions.

uturuncoglu commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -311,20 +302,15 @@ def setup_build_environment(self, env):
-            # Disables PIO-dependent code.
-            env.set("ESMF_PIO", "OFF")
+
+        # Allow override of ESMF_PIO:
```
How this works? Do we need to set `esmf_pio` in the spec if we want to use internal PIO?

theurich replied:
Not for anything newer than v8.3.1, it is automatic when you select `~external-parallelio`.

jedwards4b replied:
@uturuncoglu @theurich I feel pretty strongly that newer versions of ESMF should **never** use the internal build of PIO in spack.   It is too easy when using ESMF with CESM (for example) to end up with two different PIO versions which leads to a great deal of confusion.   Since spack easily handles the external pio build why would you ever want it to be internal?

climbfuji replied:
I agree with @jedwards4b

theurich replied:
We want to use Spack also for regression testing by the ESMF core team. In fact @uturuncoglu has started setting up GitHub actions to do this. For this we want to control turning off the default `+external-parallelio`.

jedwards4b replied:
Why not build the parallelio external for the github action as well?  That would allow you to cache the pio build and get a faster turnaround on the already very slow esmf build process.

theurich replied:
Yes, of course, we test that. But @uturuncoglu would have more details on that. But we _also_ want to test ESMF_PIO=internal. Therefore we need to be able to set `~external-parallelio`. We just want the _option_, we are not discussing the default.

uturuncoglu reviewed (dismissed):
It looks good to me. Thanks for all those changes. Do you want me to test this with GitHub action?

theurich commented:
@uturuncoglu - thanks for looking this over. I am still struggling with some formatting issues that prechecks/style does not like. Let me try to get that under control before you start testing. I will let you know.

theurich committed:
Now trying to satisfy "black reformatting".

theurich committed:
For "black" formatting really put that ugly comma at the end before
closing parentheses. Interesting.

theurich commented:
@uturuncoglu - This is ready for being tested under your GitHub action.

theurich commented:
@ jedwards4b, @climbfuji, and @AlexanderRichert-NOAA, I have no idea why your review request was automatically removed when I re-requested @uturuncoglu review. I also cannot figure out how to request review from you all again. Does anybody know? I already converted to draft, and back to open, hoping the bot would add you back on this PR.

AlexanderRichert-NOAA commented:
@theurich Not sure, but in any case I seem to have managed to re-add myself. I'm testing on a couple systems right now so I should have a verdict later today or early tomorrow.

climbfuji commented:
I added myself again. I can test next week, is that too late?

uturuncoglu commented:
@theurich how can I build ESMF with internal PIO. Is it enough to remove external one such as `esmf@8.4.1~external-parallelio`. In this case it sets `esmf_pio=auto`. Is it enough? How can you set esmf_pio as internal. I tried `esmf@8.4.1~external-parallelio esmf_pio=internal` but still sets to auto. Maybe I am using wrong syntax in here.

jedwards4b commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -14,59 +14,67 @@ class Esmf(MakefilePackage):
+    variant("xerces", default=False, description="Build with Xerces support")
     variant(
-        "parallelio",
+        "external-parallelio",
```
I do not think it wise to support the internal pio build in newer versions of esmf where the external pio is available in spack and can be built as a dependancy.

theurich replied:
@jedwards4b - I agree that 99% of Spack builds would/should use the external ParallelIO dependency. That is why it is the default under ESMF's Spack package.py. However, the option to use internal should still be there for those that want to experiment with it (e.g. @uturuncoglu ).

theurich replied:
There is a more current thread discussing exactly the same question.. resolve this one to have all the conversation on the other.

jedwards4b commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -14,59 +14,67 @@ class Esmf(MakefilePackage):
         default=True,
         description="Build with external parallelio library",
-        when="@8.3:",
+        when="@8.3.1:+mpi",
```
Parallelio~mpi works fine, why was this changed to have the +mpi?

theurich replied:
ESMF currently only supports PIO internal or external in MPI mode. This isn't a ParallelIO limitation, but an ESMF limitation.

jedwards4b replied:
It's actually not true - I build without mpi but include pio on the gust system.  Here is the spack spec:
```
esmf@8.5.0b17%intel@2021.7.1+debug+external-lapack~mpi+parallelio~xerces esmf_os=Unicos
    ^cray-libsci
    ^netcdf-c+dap~mpi
    ^parallelio@2.5.10~mpi~pnetcdf build_type=Debug
```

theurich replied:
@jedwards4b - can you point me to the resulting `esmf.mk` for this built? I am pretty sure it will have PIO turned OFF, but would like to verify.

jedwards4b replied:
/glade/u/apps/cseg/spack/opt/spack/linux-sles15-zen3/intel-2021.7.1/esmf-8.4.1b02-mdffrgnov7swkh7osc7w74fg3c7ed4vn/lib/esmf.mk
```
#
# !!! The following options were used on this ESMF build !!!
#
# ESMF_DIR: /glade/gust/scratch/jedwards/spack-stage/spack-stage-esmf-8.4.1b02-mdffrgnov7swkh7osc7w74fg3c7ed4vn/spack-src
# ESMF_OS: Linux
# ESMF_MACHINE: x86_64
# ESMF_ABI: 64
# ESMF_COMPILER: intel
# ESMF_BOPT: O
# ESMF_COMM: mpiuni
# ESMF_SITE: default
# ESMF_PTHREADS: ON
# ESMF_OPENMP: ON
# ESMF_OPENACC: OFF
# ESMF_ARRAY_LITE: FALSE
# ESMF_NO_INTEGER_1_BYTE: TRUE
# ESMF_NO_INTEGER_2_BYTE: TRUE
# ESMF_FORTRANSYMBOLS: default
# ESMF_MAPPER_BUILD: OFF
# ESMF_AUTO_LIB_BUILD: ON
# ESMF_DEFER_LIB_BUILD: ON
# ESMF_SHARED_LIB_BUILD: ON
# 
# ESMF environment variables pointing to 3rd party software:
# ESMF_MOAB:              internal
# ESMF_LAPACK:            system
# ESMF_LAPACK_LIBS:       -lsci_intel
# ESMF_ACC_SOFTWARE_STACK:            none
# ESMF_NETCDF:            nc-config
# ESMF_NETCDF_INCLUDE:    /glade/work/csgteam/spack-deployments/gust/22.12/envs/build/opt/__spack_path_placeholder__/__spack_path_placeholder__/__spack_pa/netcdf/4.9.0/packages/netcdf-c/4.9.0/intel/2021.7.1/include
# ESMF_NETCDF_LIBS:       -lnetcdf
# ESMF_NETCDF_LIBPATH:    /glade/work/csgteam/spack-deployments/gust/22.12/envs/build/opt/__spack_path_placeholder__/__spack_path_placeholder__/__spack_pa/netcdf/4.9.0/packages/netcdf-c/4.9.0/intel/2021.7.1/lib
# ESMF_NFCONFIG:          nf-config
# ESMF_NETCDFF_INCLUDE:   /glade/u/apps/gust/22.12/spack/opt/spack/netcdf/4.9.0/packages/netcdf-fortran/4.6.0/intel/2021.7.1/include
# ESMF_NETCDFF_LIBS:      -lnetcdff -lnetcdf -lnetcdf
# ESMF_NETCDFF_LIBPATH:   /glade/u/apps/gust/22.12/spack/opt/spack/netcdf/4.9.0/packages/netcdf-fortran/4.6.0/intel/2021.7.1/lib
# ESMF_PIO:               external
# ESMF_PIO_INCLUDE:       /glade/u/apps/cseg/spack/opt/spack/linux-sles15-zen3/intel-2021.7.1/parallelio-2.5.10-o56vie7zy3jbqxshdcgg5p6rzksgd2by/include
# ESMF_PIO_LIBS:          -lpioc
# ESMF_PIO_LIBPATH:       /glade/u/apps/cseg/spack/opt/spack/linux-sles15-zen3/intel-2021.7.1/parallelio-2.5.10-o56vie7zy3jbqxshdcgg5p6rzksgd2by/lib
# ESMF_YAMLCPP:           internal
```

theurich replied:
Good. Have you built an executable with it, and it runs successfully?

jedwards4b replied:
Yes, /glade/gust/scratch/jedwards/SMS_Mmpi-serial.f19_g17.X.gust_gnu.20230316_141050_qtdhib

theurich replied:
Well, that is great to know, and slightly unexpected. I did not think this would end so well because there would be a mix of mpi stub libraries going on. Therefore, there are MPI_ symbols built into the libesmf.a/.so, and I assumed that would conflict when linking (or maybe when loading at run-time) the serial-mpi that ParallelIO uses.

Apparently not a problem after all, at least not on Gust judging from your finding.

I will revert the change to keep supporting +external-parallelio~mpi.

theurich replied:
@jedwards4b One more question here. In your `esmf@8.5.0b17%intel@2021.7.1+debug+external-lapack~mpi+parallelio~xerces esmf_os=Unicos` example of the current behavior, I assume Spack found `^parallelio@2.5.10~mpi~pnetcdf build_type=Debug` because it was already there? If it was not there, it would have probably built `parallelio@2.5.10+mpi`? That is what I am seeing with the current Spack develop branch. I.e. the `~mpi` option given to `esmf` is not propagated down to `parallelio` automatically if it needs yet to be built. 

Do you confirm that, and do you see an issue with that?

jedwards4b replied:
I confirm that - the spec had to be defined very explicitly - this is a feature or limitation of spack.

theurich replied:
So I think I have that working now (see https://github.com/spack/spack/commit/e82ed24a1fa8c3c6044c44db68d0a6fea7c8b46d), where the `+mpi/~mpi` variant is passed down to the `parallelio` dependency. I had some issue getting things build all the way through locally until I made a small change in my local `netcdf-c/package.py`, basically again handing the `+mpi/~mpi` through to `hdf5`, or else it was trying to pick a `+mpi` variant, but then failed linking against it during the `parallelio` install. But that's outside of ESMF, and I guess who ever is using `~mpi` with `+external-parallel` knows what they are doing.

jedwards4b reviewed (dismissed):
It looks like you've left the external-parallelio as the default so I will accept this change
if you feel strongly about it.   But I don't see any reason to ever use the internal pio.

theurich commented:
> I added myself again. I can test next week, is that too late?

@climbfuji the more tests the better. Even next week. :-)

theurich commented:
@spackbot run pipeline

uturuncoglu commented:
@theurich I could able to install ESMF with external and internal PIO without any issue under GitHub action (just GNU). I did not run tests but maybe I could implement way to run NUOPC proto apps.

theurich committed:
Merge branch 'develop' into esmf-update

theurich commented:
@AlexanderRichert-NOAA @climbfuji @jedwards4b @uturuncoglu - I think this PR needs another round of approvals. But probably #35969 should go first? And then we come back to this one? There shouldn't be any conflict, so I think it's okay. Any thoughts?

AlexanderRichert-NOAA commented:
I don't know if the order matters? Either way someone will have to merge changes. BTW there's also #34579.

uturuncoglu commented:
It looks good to me

climbfuji reviewed (dismissed):
Approved, but still agreeing with @jedwards4b that internal PIO should not be an option for any newer version of esmf that supports external PIO. However, I am not the main developer of the package.

theurich committed:
Merge branch 'develop' into esmf-update

theurich commented:
@tldahlgren - could you look at this PR as a follow-on to #35969 that was just merged? Thank you!

theurich committed:
Support building against external-parallelio even w/o mpi, but select
the external-parallelio dependency accordingly.

theurich committed:
Correct C compiler setting.

uturuncoglu commented:
@theurich after last changes I could not compile ESMF without pnetcdf. As you know that I had issue with PIO when ESMF is build with pnetcdf under NoahMP. I could not look at that one more deeply but I wonder if there is a special limitation in ESMf side about building it without parallel-netcdf. To be more specific, I am getting following error from spack

```
  ==> Error: Cannot set variant 'pnetcdf' for package 'esmf' because the variant condition cannot be satisfied for the given spec
```

and I was using following to build ESMF - `esmf@8.4.1~pnetcdf` without any issue. Any suggestion?

theurich commented:
@uturuncoglu - are you saying that just `esmf@8.4.1~pnetcdf` is giving you this error? That would be strange. The only reason I see why `pnetcdf` would not be a valid variant is if you had also specified `~mpi`.

uturuncoglu commented:
@theurich let me test it also in an isolated environment like my docker container. This error is coming from Github action and it was working with your branch before. Maybe a recent commit is causing the issue.

theurich commented:
> @theurich let me test it also in an isolated environment like my docker container. This error is coming from Github action and it was working with your branch before. Maybe a recent commit is causing the issue.

Can you point me to the GitHub action error log?

uturuncoglu commented:
@theurich It in here, https://github.com/uturuncoglu/esmf/actions/runs/4444450575/jobs/7802665771
just took at "Install ESMf with Spack" step. It is also shown passed but actually failed. I am still working on this action.

theurich commented:
@ufu

> @theurich It in here, https://github.com/uturuncoglu/esmf/actions/runs/4444450575/jobs/7802665771 just took at "Install ESMf with Spack" step. It is also shown passed but actually failed. I am still working on this action.

@uturuncoglu - I wonder if somehow it could not find an MPI and decided to implicitly set `~mpi`. If that happens, you cannot refer to `pnetcdf`, not even to say `~pnetcdf`. Maybe this should be handled differently?

uturuncoglu commented:
@theurich let me test with gnu and openmpi since intel could have issue related with MPI.

uturuncoglu commented:
@theurich it works fine under GNU. Maybe you are right about the MPI issue under Intel oneAPI.

theurich commented:
> @theurich it works fine under GNU. Maybe you are right about the MPI issue under Intel oneAPI.

@uturuncoglu Okay, that makes the most sense to me then from what we know so far. With that in mind, I think it will be better to change `esmf/package.py` so that `pnetcdf` is always a valid variant. And default is `False`. This is also what I see `ParallelIO` is doing, and it makes a lot more sense to me that way because it work for `+mpi` and `~mpi` alike.

I will make the change here, and then it can get reviewed (and retested in your actions) again. As a side effect, you will not have to explicitly disable `pnetcdf`, because it will default to `False`. But if you do, it won't hurt either... which is more robust.

uturuncoglu commented:
@theurich Sure. That sounds fine to me. Once it is ready I could test it again. I still need time to make action work with oneAPI but I think I am very close.

theurich committed:
Handle `pnetcdf` variant consistent with how `ParallelIO` does it. And
also pass the `pnetcdf` variant down to the `external-parallelio`
dependency if set.

theurich commented:
@uturuncoglu It's ready for you to test again. The `pnetcdf` variant is always available now, regardless of `mpi`. Default is `False`. Plus it gets passed down to the `external-parallelio` dependency if set. 
Let me know if you run into anything else. Thanks!

theurich committed:
Long line formatting again.

theurich commented:
@spackbot run pipeline

theurich commented:
@uturuncoglu @AlexanderRichert-NOAA @climbfuji @jedwards4b -  I was out all of last week, but back at work again today. I'd like to move this PR forward if possible.

I know there was a preference expressed to not at all offer the option to build the internal version of PIO as a Spack variant of ESMF. At this point though the ESMF core team would like to have access to this option even through Spack for testing purposes. The `external-parallelio` is the default variant, and a user explicitly has to negate it to end up with the internal version, so I think it is pretty safe as to not having user's accidentally end up with the internal PIO build.

Anyway, I think this PR needs a review from everyone assigned once more before it can be merged. 

@tldahlgren, would you be available to see this one through once the reviews are in? Thank you.

jedwards4b commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -14,59 +14,67 @@ class Esmf(MakefilePackage):
     variant("netcdf", default=True, description="Build with NetCDF support")
-    variant("pnetcdf", default=True, description="Build with pNetCDF support", when="+mpi")
-    variant("xerces", default=True, description="Build with Xerces support")
+    variant("pnetcdf", default=False, description="Build with pNetCDF support")
```
Why should this default to false?

theurich replied:
I guess for the same reason that it defaults to `False` for the ParallelIO package. Basically I aligned it with that... which also takes care of the issue that @uturuncoglu had reported.

jedwards4b replied:
I see, thanks -- I forgot about that.

jedwards4b reviewed (changes requested):
I think that this is out of date with respect to develop which we updated last week in your absence - we (@climbfuji ) also found that an additional flag was needed on Narwhal.  
I've tested this on gust with no negative affects, I think you should add that change to this PR:  https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555

theurich committed:
Merge branch 'develop' into esmf-update

theurich commented:
> I think that this is out of date with respect to develop which we updated last week in your absence - we (@climbfuji ) also found that an additional flag was needed on Narwhal. I've tested this on gust with no negative affects, I think you should add that change to this PR: [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555)

@jedwards4b - I just merged the updated `develop` into this branch... so should be good now. The thing on Narwhal looks strange to me, and I think it should be dealt with as a separate issue. I already commented on https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555. But what ever we come out with I don't think should be wrapped into this PR that's already pretty big, and been hanging around for quite some time.

climbfuji commented:
> > I think that this is out of date with respect to develop which we updated last week in your absence - we (@climbfuji ) also found that an additional flag was needed on Narwhal. I've tested this on gust with no negative affects, I think you should add that change to this PR: [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555)
> 
> @jedwards4b - I just merged the updated `develop` into this branch... so should be good now. The thing on Narwhal looks strange to me, and I think it should be dealt with as a separate issue. I already commented on [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555). But what ever we come out with I don't think should be wrapped into this PR that's already pretty big, and been hanging around for quite some time.

I agree. The Narwhal issue might be part of a bigger problem. Maybe it has something to do with how Cray systems are treated in spack. The authoritative repo moved from the compiler wrappers to the direct compiler calls on newer crays (depending on some version of some cray library, auto-detected.) But we did have this change from spack in our fork for the last release, and therefore also for the successful build of 8.3.0b09 before the "edit" -> "setup_build_environment" change was made. There is something going on under the hood ...

climbfuji commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -15,60 +15,67 @@ class Esmf(MakefilePackage):
+        sha256="0ff43ede83d1ac6beabd3d5e2a646f7574174b28a48d1b9f2c318a054ba268fd",
+        deprecated=True,
+    )
+    version("8.3.0b09", commit="5b7e546c4b", deprecated=True)
```
Note that his is still the version used by the ufs-weather-model develop branch as of today. We shouldn't deprecate it until the UFS WM moved away (because we need to install it everywhere as part of our spack-stack releases, and I don't want to add the flag to install deprecate packages as a blanket for the entire environment).

theurich replied:
@climbfuji I see. I had thought that ufs-weather-model develop branch had moved on to ESMF 8.4.1 now -  @junwang-noaa can you help clarify?
My rationale for deprecation marking in this PR was that if there was a patch release, e.g. 8.3.1, then mark 8.3.0 deprecated (same for 8.4.1 over 8.4.0) simply because if there is a patch, it meant that there was actually an issue found in the .0 release that needed fixing. Therefore marked deprecated.
But I also understand your concern with regards to downstream changes needed if ufs-weather-model is right at 8.3.0.
Let's clarify this, and if the impact is unreasonable, then remove the `deprecated=True` from the version that you need for continued support.

junwang-noaa replied:
@theurich  We have a ufs-weather-model branch with ESMF 8.4.1 tested. We are waiting for the libraries to be installed on wcoss2 before we can commit the code.

theurich replied:
Thanks @junwang-noaa. So @climbfuji, it looks like this might come down to a timing question. Jun, how close do you think we are with v8.4.1 as the official ESMF dependency for ufs-weather-model? A matter of days, or more like weeks?
Also, just to clarify, it is actually v8.3.0b09 that is currently the official dependency, right? Definitely be great to get away from a beta tag, and mark it as deprecated when we can... but I am okay to have it as a follow-on PR if timing is better that way. Better than holding up this PR.

climbfuji replied:
It’s a bit of a chicken-and-egg problem. esmf 8.4.1 will be made available with spack-stack, but in order to switch to spack-stack it is easier to have 8.3.0b09. However, if you make the change now it won’t go into our release branch (but come to develop later), so maybe after all just leave it like it is.

And maybe deprecating that version will speed up the update in the ufs-weather-model!

climbfuji commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -207,9 +241,11 @@ def setup_build_environment(self, env):
 
         if "+mpi" in spec:
             env.set("ESMF_CXX", spec["mpi"].mpicxx)
+            env.set("ESMF_C", spec["mpi"].mpicc)
```
Maybe this helps with the problems on Narwhal ...

theurich replied:
I doubt it will. This is just to make sure the C compiler is set up correctly ... which is important for the ESMF develop branch where we make a cleaner distinction between code that's pure C vs C++. It should not affect the C++ compilation/linking that was the issue on your latest Narwhal Spack builds. But it also should not hurt.

tldahlgren commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -15,60 +15,67 @@ class Esmf(MakefilePackage):
+    )
+    version(
+        "7.1.0r",
+        sha256="ae9a5edb8d40ae97a35cbd4bd00b77061f995c77c43d36334dbb95c18b00a889",
```
I am not able to confirm this sha256.  I get `Could not find any remote versions for ESMF` when I run `spack checksum esmf@7.1.0r`.  If I add `url="https://github.com/esmf-org/esmf/archive/refs/tags/ESMF_7_1_0r.tar.gz,"` to the `7.1.0r` `version` directive, the sha256 is `e08f21544083dcbe162b472852e321f8df14f4f711f35508403d32df438367a7`.

theurich replied:
Thanks for checking. I simplified the tarball URL construction, and all tarballs now from GitHub (not sure why this was done differently before). Some checksums needed updating... should all be consistent now.

tldahlgren commented on var/spack/repos/builtin/packages/esmf/package.py:
```diff
@@ -249,12 +289,10 @@ def setup_build_environment(self, env):
-                # The mpich 3 series split apart the Fortran and C bindings,
-                # so we link the Fortran libraries when building C programs:
-                env.set("ESMF_CXXLINKLIBS", "-lmpifort")
+                if self.version < Version("8.3.0"):
```
When checking the version of the current package, `self.spec.satisfies()` should be used (see https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html#querying-spec-versions).

theurich replied:
Thanks for the tip. Two places were updated to follow the recommendation.

theurich committed:
Simplify handling of tarball URL construction and update sha256
checksums.

theurich committed:
Align version check with recommended self.spec.satisfies().

theurich committed:
Deprecate v8.4.0 which has a bug that can cause memory corruption, fixed
in v8.4.1.

theurich committed:
Use double quotes vs single quotes as per style-check... although
https://spack-tutorial.readthedocs.io/en/latest/tutorial_packaging.html#querying-spec-versions
clearly shows it with single quotes.

theurich commented:
> > > I think that this is out of date with respect to develop which we updated last week in your absence - we (@climbfuji ) also found that an additional flag was needed on Narwhal. I've tested this on gust with no negative affects, I think you should add that change to this PR: [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555)
> > 
> > 
> > @jedwards4b - I just merged the updated `develop` into this branch... so should be good now. The thing on Narwhal looks strange to me, and I think it should be dealt with as a separate issue. I already commented on [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555). But what ever we come out with I don't think should be wrapped into this PR that's already pretty big, and been hanging around for quite some time.
> 
> I agree. The Narwhal issue might be part of a bigger problem. Maybe it has something to do with how Cray systems are treated in spack. The authoritative repo moved from the compiler wrappers to the direct compiler calls on newer crays (depending on some version of some cray library, auto-detected.) But we did have this change from spack in our fork for the last release, and therefore also for the successful build of 8.3.0b09 before the "edit" -> "setup_build_environment" change was made. There is something going on under the hood ...

@climbfuji it does seem problematic to me to have moved away from the Cray compiler wrappers under Spack. There is a good reason they provide those on the Crays, and the ESMF build system relies on them providing the correct compiler and linker flags... so we don't explicitly deal with it. This is especially important because ESMF mixes Fortran, C, and C++, so linking can be a pain.

tldahlgren reviewed (approved):
Confirmed all eleven version sha256.

climbfuji commented:
> > > > I think that this is out of date with respect to develop which we updated last week in your absence - we (@climbfuji ) also found that an additional flag was needed on Narwhal. I've tested this on gust with no negative affects, I think you should add that change to this PR: [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555)
> > > 
> > > 
> > > @jedwards4b - I just merged the updated `develop` into this branch... so should be good now. The thing on Narwhal looks strange to me, and I think it should be dealt with as a separate issue. I already commented on [NOAA-EMC/spack-stack#517 (comment)](https://github.com/NOAA-EMC/spack-stack/issues/517#issuecomment-1483548555). But what ever we come out with I don't think should be wrapped into this PR that's already pretty big, and been hanging around for quite some time.
> > 
> > 
> > I agree. The Narwhal issue might be part of a bigger problem. Maybe it has something to do with how Cray systems are treated in spack. The authoritative repo moved from the compiler wrappers to the direct compiler calls on newer crays (depending on some version of some cray library, auto-detected.) But we did have this change from spack in our fork for the last release, and therefore also for the successful build of 8.3.0b09 before the "edit" -> "setup_build_environment" change was made. There is something going on under the hood ...
> 
> @climbfuji it does seem problematic to me to have moved away from the Cray compiler wrappers under Spack. There is a good reason they provide those on the Crays, and the ESMF build system relies on them providing the correct compiler and linker flags... so we don't explicitly deal with it. This is especially important because ESMF mixes Fortran, C, and C++, so linking can be a pain.

This change was made by the spack developers based on the fact that newer crays move away from the compiler wrappers (apparently). Entirely outside my purview. I've made sure, where possible, to use the wrappers on the existing systems that by coincidence have the newer library that spack looks for, but were designed to use the wrappers.

uturuncoglu commented:
@climbfuji @theurich you could pass spack `--deprecated` to spack install command. So, mark as deprecated is not important I think. It is just an extra information and experienced users could still install those.

climbfuji commented:
> @climbfuji @theurich you could pass `--deprecated` argument to spack install command. So, mark as deprecated is not important I think. It is just an extra information and experienced users could still install those.

To be clear, we install about 200-ish packages per compiler in our environments. This must be achieved with a single `spack install` command. If we pass `--deprecated` for ESMF, then it is a blanket cover for any deprecated package, which we don't want. But as I said above, the timing of this change is such that it shouldn't affect us. By the time this change comes to the spack-stack fork, the ufs-weather-model will have switched to 8.4.1.

theurich commented:
@tldahlgren - It looks like we're all cleared to move this one forward. Thank you!

tldahlgren commented:
I noticed the comment about the docs not being up-to-date w.r.t. the newer style checks (e.g., single versus double quotes).  We're slowly updating them when we get in there to change things.
